daemon/graphdriver/copy/copy.go

```diff
@@ -11,20 +11,32 @@ package copy // import "github.com/docker/docker/daemon/graphdriver/copy"
 */
 import "C"
 import (
-	"container/list"
+	"context"
 	"fmt"
 	"io"
+	"math/rand"
 	"os"
 	"path/filepath"
+	"sync"
 	"syscall"
 	"time"
 
 	"github.com/docker/docker/pkg/pools"
 	"github.com/docker/docker/pkg/system"
 	rsystem "github.com/opencontainers/runc/libcontainer/system"
+	"golang.org/x/sync/errgroup"
 	"golang.org/x/sys/unix"
 )
 
+const (
+	// Docker sets this earlier on in initialization to 0022
+	// Unfortunately, sometimes the special bits are not respected
+	// and we have to force-set them.
+	desiredUmask = 7022
+	// This is based on some empirical testing on EXT4 filesystems
+	workerCount = 8
```

Comment:
Maybe this should be configurable via --storage-opt?

Reply:
Do you foresee negative side-effects to this change? i.e.; is there a specific reason you want this to be configurable @AkihiroSuda ?

I'm not _against_ making things tuneable, but also want to guard against adding an option for _everything_ (also because work is in progress to delegate more things to containerd)

Reply:
Wondering high concurrencycan be problematic for poor machines, but no benchmark data

Reply:
Wondering high concurrencycan be problematic for poor machines, but no benchmark data

Reply:
Perhaps `workerCount` should be capped to number of CPUs on the machine?

Reply:
Number of CPU's gives worse performance (unless of course nCPU > 8)... on a single core system is WAY worse
 
This is from the branch head compared to setting the worker count to the number of CPU's on a single core system
```
benchmark                              old ns/op     new ns/op     delta
BenchmarkCopyDirFewerDirsMoreFiles     340916        4474857       +1212.60%
```

This issue here is this is all I/O and not CPU bound and so the runtime is just waiting there are too few workers.

Reply:
So, at least on XFS, setting workers to more than 8 has degradation. Setting it to 64 (our CPU count) hurts pretty bad.

Reply:
In this case do you think it's worthwhile to add a storage-opt for vfs?

Reply:
Oh yeah and just a note, I ran these benches using ext4.

Reply:
It sounds like giving the user this knob would be counter productive, except for low memory machines. I would suggest making the channel depth, and worker count environment variables, if we really want to.

Reply:
Or a storage-opt? Could be nice for people with beefy machines as well.

As I recall, a single worker was *much* slower with this change as compared to the original.